opencog/atomspace/Foreach.h

```diff
@@ -18,16 +18,16 @@ namespace opencog
  */
 
 
-//! Invoke the callback on each atom in the outgoing set of handle h.
+//! Invoke the callback on each atom in the outgoing set of handle h
+//! till one of them return true, and returns true as a
+//! result. Otherwise the callback is called on all outgoings and
+//! false is returned.
 /// XXX TODO this should be moved to class Link
 template<class T>
 inline bool foreach_outgoing_handle(LinkPtr ll, bool (T::*cb)(Handle), T *data)
 {
-    const std::vector<Handle> &vh = ll->getOutgoingSet();
-    size_t sz = vh.size();
-
-    for (size_t i = 0; i < sz; i++) {
-        bool rc = (data->*cb)(vh[i]);
+    for (Handle out_h : ll->getOutgoingSet()) {
+        bool rc = (data->*cb)(out_h);
         if (rc) return rc;
```

Comment:
One reason I often used the old style, of putting the size in an int, and looping on an int, is that I beleive that the compiler will turn this into efficient.(at least, w.r.t. the loop termination condition).  By contrast, your new code is not obviously as efficient. I can't tell what the compiler would actually do in this case.   Note that this is fairly performance-critical code: the entire pattern matcher is a recursive nesting of these for-loops.   So I agonize about this ...